src/tracer/MOM_neutral_diffusion.F90

```diff
@@ -877,16 +946,292 @@ subroutine find_neutral_surface_positions(nk, Pl, Tl, Sl, dRdTl, dRdSl, Pr, Tr,
       hL = absolute_position(nk,Pl,KoL,PoL,k_surface) - absolute_position(nk,Pl,KoL,PoL,k_surface-1)
       hR = absolute_position(nk,Pr,KoR,PoR,k_surface) - absolute_position(nk,Pr,KoR,PoR,k_surface-1)
       if ( hL + hR > 0.) then
-        hEff(k_surface-1) = 2. * hL * hR / ( hL + hR )
+        hEff(k_surface-1) = 2. * hL * hR / ( hL + hR ) ! Harmonic mean of layer thicknesses
       else
         hEff(k_surface-1) = 0.
       endif
     endif
 
   enddo neutral_surfaces
 
-end subroutine find_neutral_surface_positions
+end subroutine find_neutral_surface_positions_continuous
+
+!> Higher order version of find_neutral_surface_positions. Returns positions within left/right columns
+!! of combined interfaces using intracell reconstructions of T/S
+subroutine find_neutral_surface_positions_discontinuous(nk, Pres_l, Tl, Sl, dRdT_l, dRdS_l, &
```

Comment:
I spent a bit of time trying to shoehorn your original algorithm into the discontinuous one, but to no avail. What it boiled down to was that the original algorithm relies on the fact that the layer index always corresponds to the top interface. With the discontinuous reconstructions a different mapping between interface and layer exists, plus you may potentially need to search across the discontinuity.

I did do something similar to the test you suggested when I was prototyping the logic for the new one for some really simple cases that produced qualitatively similar results (didn't check for bitwise reproduction). If they do, then I agree that the code could be collapsed so that there's not a huge bit of similar code.